src/config.coffee

```diff
@@ -1,41 +1,349 @@
-	configMountPoint: checkString(process.env.CONFIG_MOUNT_POINT) ? '/mnt/mmcblk0p1/config.json'
-	ledFile: checkString(process.env.LED_FILE) ? '/sys/class/leds/led0/brightness'
-	bootstrapRetryDelay: checkInt(process.env.BOOTSTRAP_RETRY_DELAY_MS, positive: true) ? 30000
-	restartSuccessTimeout: checkInt(process.env.RESTART_SUCCESS_TIMEOUT, positive: true) ? 60000
-	appUpdatePollInterval: checkInt(process.env.APPLICATION_UPDATE_POLL_INTERVAL, positive: true) ? 60000
-	successMessage: 'SUPERVISOR OK'
-	forceSecret:
-		api: checkString(process.env.RESIN_SUPERVISOR_SECRET) ? null
-		logsChannel: checkString(process.env.RESIN_SUPERVISOR_LOGS_CHANNEL) ? null
-	vpnStatusPath: checkString(process.env.VPN_STATUS_PATH) ? '/mnt/root/run/openvpn/vpn_status'
-	hostOSVersionPath: checkString(process.env.HOST_OS_VERSION_PATH) ? '/mnt/root/etc/os-release'
-	dockerRoot: dockerRoot
-	btrfsRoot: checkString(process.env.BTRFS_ROOT) ? "#{dockerRoot}/btrfs/subvolumes"
-	privateAppEnvVars: [
-		'RESIN_SUPERVISOR_API_KEY'
-		'RESIN_API_KEY'
-	]
-	dataPath: checkString(process.env.RESIN_DATA_PATH) ? '/resin-data'
-	bootMountPoint: checkString(process.env.BOOT_MOUNTPOINT) ? '/boot'
-	proxyvisorHookReceiver: checkString(process.env.RESIN_PROXYVISOR_HOOK_RECEIVER) ? 'http://0.0.0.0:1337'
+Promise = require 'bluebird'
+_ = require 'lodash'
+Lock = require 'rwlock'
+deviceRegister = require 'resin-register-device'
+fs = Promise.promisifyAll(require('fs'))
+EventEmitter = require 'events'
+
+{ writeAndSyncFile, writeFileAtomic } = require './lib/fs-utils'
+osRelease = require './lib/os-release'
+supervisorVersion = require './lib/supervisor-version'
+constants = require './lib/constants'
+
+module.exports = class Config extends EventEmitter
+	constructor: ({ @db, @configPath }) ->
+		# These are values that come from env vars or hardcoded defaults and can be resolved synchronously
+		# Defaults needed for both gosuper and node supervisor are declared in entry.sh
+		@constants = constants
+
+		@funcs =
+			version: ->
+				Promise.resolve(supervisorVersion)
+			currentApiKey: =>
+				@getMany([ 'apiKey', 'deviceApiKey' ])
+				.then ({ apiKey, deviceApiKey }) ->
+					return apiKey ? deviceApiKey
+			offlineMode: =>
+				@getMany([ 'resinApiEndpoint', 'supervisorOfflineMode' ])
+				.then ({ resinApiEndpoint, supervisorOfflineMode }) ->
+					return Boolean(supervisorOfflineMode) or !Boolean(resinApiEndpoint)
+			pubnub: =>
+				@getMany([ 'pubnubSubscribeKey', 'pubnubPublishKey' ])
+				.then ({ pubnubSubscribeKey, pubnubPublishKey }) ->
+					pubnub = {
+						subscribe_key: pubnubSubscribeKey
+						publish_key: pubnubPublishKey
+						ssl: true
+					}
+					return pubnub
+			resinApiEndpoint: =>
+				# Fall back to checking if an API endpoint was passed via env vars if there's none in config.json (legacy)
+				@get('apiEndpoint')
+				.then (apiEndpoint) ->
+					return apiEndpoint ? @constants.apiEndpointFromEnv
```

Comment:
REQUIRED CHANGE

```
@constants.apiEndpointFromEnv
```
will fail because you used a thin arrow - I think it should just be `constants.apiEndpointFromEnv` though in order to avoid this source of confusion

Reply:
Done